### PR TITLE
add ozone sonder plot based on 'develop aircraft' branch

### PR DESCRIPTION
**Pull request by btang1**
I have added paired ozone sonder, and 2 plots (vertical time series & vertical boxplot) for ozone sonder.

this branch is based on 'develop aircraft branch'

### CODE REVIEW

**Comment by btang1** [comment]
@colin-harkins @rschwant @zmoon could you please give a review on this?

and I am continuing update some new features on this branch. for example (aggregate plot & density scatter plot)

**Comment by rschwant** [comment]
@zmoon and @btang1 is it possible to have this pull request just be a PR to the develop branch with only the changes you made Beiming to add the ozonesonde analysis & plots? When I look at the changed files to do the review, it includes all the files from the entire old develop_aircraft branch, which makes it hard to review? Also can you update the code to resolve the conflicts in the driver.py file.

File: melodies_monet/driver_backup.py
Comment (zmoon) [inline melodies_monet/driver_backup.py:1]: This should most likely be removed.

File: melodies_monet/driver.py [hunk 1]
```diff
@@ -1257,6 +1301,8 @@ def plotting(self):
             from .plots import satplots as splots,savefig
         else: 
             from .plots import surfplots as splots, savefig
+            from .plots import aircraftplots as airplots
```
Comment (zmoon) [inline melodies_monet/driver.py:1304]: @rschwant seems like this line is necessary for aircraft plots to work, though it wasn't in develop

**Comment by zmoon** [comment]
@rschwant should be easier to review this now

File: melodies_monet/driver.py (hunk 1 above)
Comment (rschwant) [reply melodies_monet/driver.py:1303]: I see, it got deleted with one of the satellite updates. Thanks for noticing this.

**Comment by btang1** [comment]
Sorry for the obsolete code with aircraft, this is done at a very early age. Thanks zach for carrying me~

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        f,ax = plt.subplots(**fig_dict)
+    else:
+        f,ax = plt.subplots(figsize=(8,8))
+    #set default text size
+    def_text = dict(fontsize=20)
+    if text_dict is not None:
+        text_kwargs = {**def_text, **text_dict}
+    else:
+        text_kwargs = def_text
+
```
Comment (rschwant) [inline melodies_monet/plots/ozone_sonder_plots.py:55]: Having these plots replicated by altitude type seems redudent to me. Can you instead change this to:

if altitude_method[0] =='sea level':
      alt_p = ALT_sl
      alt_p_name = 'ALT-sea level (km)'
elif altitude_method[0] =='ground level':
      alt_p = ALT_gl
      alt_p_name = 'ALT-ground level (km)'

And then just have one plotting routine that uses these variables. You can probably even condense the calculations above into this logic too. This reduces the complexity of the code.

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+#                  'zorder':0,'showfliers': False
+#                  'showmeans': False,'notch'= False,'showfliers': False
+                  }
+    
+    #Make plot
+    #get position list
+    position_list_1 = altitude_threshold_list[:-1]
+    position_list_2 = altitude_threshold_list[1:]
+    position_list_mid = [(position_list_1[i]+position_list_2[i])/2 for i in range(len(position_list_1))]
+
```
Comment (rschwant) [inline melodies_monet/plots/ozone_sonder_plots.py:135]: Having these plots replicated by altitude type seems redudent to me. Can you instead change this to:

if altitude_method[0] =='sea level':
      alt_p = ALT_sl
      alt_p_name = 'ALT-sea level (km)'
elif altitude_method[0] =='ground level':
      alt_p = ALT_gl
      alt_p_name = 'ALT-ground level (km)'

And then just have one plotting routine that uses these variables. You can probably even condense the calculations above into this logic too. This reduces the complexity of the code.

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            ax.plot(O3_MODEL_ALL[i],ALT_sl,'-*',color=fill_color_list[i+1],label = 'MODEL:'+model_name_list[i+1])
+
+            colors=[fill_color_list[i+1]]*(len(altitude_threshold_list)-1)
+            for patch,color in zip(bplot_model['boxes'],colors):
+                patch.set_edgecolor(color)
+
+        plt.title('Comparison at '+str(station_name[0])+' on '+str(release_time)+' UTC')
+        plt.legend()
+        plt.ylim(altitude_range[0],altitude_range[1])
+        plt.xlim(ozone_range[0],ozone_range[1])
```
Comment (rschwant) [inline melodies_monet/plots/ozone_sonder_plots.py:155]: For both of these can you update the xlim to be based on the vmin_plot and vmax_plot values instead that are specified in the observation class. This keeps things consistent with the surface plots and also prepares us to use this code for the met sonde data too.

I recognize the altitude is now specified differently in all these plots we can unify this later.

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        plt.ylabel('ALT-sea level (km)')
+        plt.xlabel('O3 (ppbv)')
```
Comment (rschwant) [inline melodies_monet/plots/ozone_sonder_plots.py:157]: Once you code the vmin_plot and vmax_plot in you will also be able to access the ylabel_plot and could use that here too to make this code more universal and not just for Ozone. And you could expand your code to look at temperature for example very easily with some small updates like this.

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+         })
+    if altitude_method[0] == 'ground level':
+        height_value = df_height.loc[df_height['station']==station_name[0]].values[0][1]
+    elif altitude_method[0] == 'sea level':
+        height_value = 0
+
+    #get o3 model, o3 sonder (obs) and height
+    df_short = df[df['altitude']<altitude_range[1]+height_value]
+    ALT = df_short['altitude']-height_value
+    O3_OBS = df_short['o3']
```
Comment (rschwant) [inline melodies_monet/plots/ozone_sonder_plots.py:204]: You could add obsvar and modvar in line above and below to keep this more variable agnostic. Then you can use the same code for the met analysis too.

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    cb=plt.colorbar(sc)
+    if altitude_method[0] == 'ground level':
+        cb.set_label('Ground Level Altitude (km)',fontsize=15)
+    elif altitude_method[0] == 'sea level':
+        cb.set_label('Sea Level Altitude (km)',fontsize=15)
+
+    #add some points to make best fit go entire domain
+    slope = np.poly1d(np.polyfit(O3_OBS, O3_MODEL, 1))
+    Modify_OBS = O3_OBS.to_list()
+    Modify_OBS.append(ozone_range[0])
```
Comment (rschwant) [inline melodies_monet/plots/ozone_sonder_plots.py:218]: For these (and the references below too), can you update the ozone_range to be based on the vmin_plot and vmax_plot values instead that are specified in the observation class. This keeps things consistent with the surface plots and also prepares us to use this code for the met sonde data too.

I recognize the altitude is now specified differently in all these plots we can unify this later.

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    plt.plot(Modify_OBS, slope(Modify_OBS),color='k',linestyle='-.',label='best fit')
+   
+    #plot Y=X line
+    plt.axline((ozone_range[1],ozone_range[1]),slope=1,color='k',linestyle='-',label='Y=X')
+
+    plt.xlim(ozone_range[0],ozone_range[1])
+    plt.ylim(ozone_range[0],ozone_range[1])
+    plt.xlabel('O$_3$ Obs (ppbv)')
+    plt.ylabel('O$_3$ Model (ppbv)')
+    plt.legend()
```
Comment (rschwant) [inline melodies_monet/plots/ozone_sonder_plots.py:230]: Once you code the vmin_plot and vmax_plot in you will also be able to access the ylabel_plot and could use that here too to make this code more universal and not just for Ozone. And you could expand your code to look at temperature for example very easily with some small updates like this.

You can just use the text in ylabel_plot and then append Obs to the xlabel and append Model to the ylabel here to make this more generic.

**Review by rschwant — COMMENTED** [review]
Great job, this is looking really good! And it's great to have this capability added. I added some comments to make the code cleaner and also to use the functionality already available in the surface plots for providing the ozone min and max values and also for providing the model color types. It would be best to not have these namelist options be different for each plot as this is confusing to users. Let me know if you have questions. The part that Zach noted where the aircraft plots accidentally got deleted needs to be fixed this week because all the aircraft plots are broken now in develop. If you can address these comments this week, then it can go in with this PR. But if not no problem, I can submit a separate PR for this problem too.

**Comment by btang1** [comment]
> Great job, this is looking really good! And it's great to have this capability added. I added some comments to make the code cleaner and also to use the functionality already available in the surface plots for providing the ozone min and max values and also for providing the model color types. It would be best to not have these namelist options be different for each plot as this is confusing to users. Let me know if you have questions. The part that Zach noted where the aircraft plots accidentally got deleted needs to be fixed this week because all the aircraft plots are broken now in develop. If you can address these comments this week, then it can go in with this PR. But if not no problem, I can submit a separate PR for this problem too.

thanks @rschwant I will try to address the changes (very helpful) and have a test run. I will let you know the results. if it works, I will update the PR. and then we can merge it.

File: melodies_monet/driver.py [hunk 9]
```diff
@@ -1291,8 +1336,44 @@ def plotting(self):
                 region_name = grp_dict['region_name'] 
                 region_list = grp_dict['region_list']
                 model_name_list = grp_dict['model_name_list']     
-            
 
+            #read-in special settings for vertical_single_date
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/driver.py:1340]: This part is done

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+    #set default text size
+    def_text = dict(fontsize=20)
+    if text_dict is not None:
+        text_kwargs = {**def_text, **text_dict}
+    else:
+        text_kwargs = def_text
+
+    #Make plot
+    if altitude_method[0] =='sea level':
+        ax.plot(O3_OBS,ALT_sl,'-*',color=fill_color_list[0],label = 'OBS:'+model_name_list[0])
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:58]: this part is done!

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        for i in range(len(O3_MODEL_ALL)): 
+            ax.plot(O3_MODEL_ALL[i],ALT_sl,'-*',color=fill_color_list[i+1],label = 'MOD:'+model_name_list[i+1])
+        plt.title('Comparison at '+str(station_name[0])+' on '+str(release_time)+' UTC')
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:61]: this part is done

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        ax.set_ylabel('ALT-sea level (km)')
+        ax.set_xlabel('O3 (ppbv)')
+    elif altitude_method[0] == 'ground level':
+        ax.plot(O3_OBS,ALT_gl,'-*',color=fill_color_list[0],label = 'OBS:'+model_name_list[0])
+        for i in range(len(O3_MODEL_ALL)):
+            ax.plot(O3_MODEL_ALL[i],ALT_gl, '-*',color=fill_color_list[i+1],label = 'MOD:'+model_name_list[i+1])     
+        plt.title('Comparison at '+str(station_name[0])+' on '+str(release_time)+' UTC')
+        plt.legend()
+        plt.ylim(altitude_range[0],altitude_range[1])
+        plt.xlim(ozone_range[0],ozone_range[1])
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:74]: this part it done!

File: melodies_monet/plots/ozone_sonder_plots.py (hunk 2 above)
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:55]: this part is done!

File: melodies_monet/plots/ozone_sonder_plots.py (hunk 3 above)
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:135]: this part is resolved

File: melodies_monet/plots/ozone_sonder_plots.py (hunk 4 above)
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:155]: this section is done!

File: melodies_monet/plots/ozone_sonder_plots.py (hunk 6 above)
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:204]: this is done

File: melodies_monet/plots/ozone_sonder_plots.py (hunk 7 above)
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:218]: This is done!

File: melodies_monet/driver.py [hunk 13]
```diff
@@ -1291,8 +1336,44 @@ def plotting(self):
...
+            if plot_type == 'vertical_single_date':
+                model_name_list = grp_dict['model_name_list']
+                altitude_range = grp_dict['altitude_range']
+                altitude_method = grp_dict['altitude_method']
+                ozone_range = grp_dict['ozone_range']
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/driver.py:1345]: this is done!

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
+# Copyright (C) 2022 National Center for Atmospheric Research and National Oceanic and Atmospheric Administration
+# SPDX-License-Identifier: Apache-2.0
+#
+import os
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:4]: this is done!

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+    elif altitude_method[0] == 'ground level':
+        ax.plot(O3_OBS,ALT_gl,'-*',color=fill_color_list[0],label = 'OBS:'+model_name_list[0])
+        for i in range(len(O3_MODEL_ALL)):
+            ax.plot(O3_MODEL_ALL[i],ALT_gl, '-*',color=fill_color_list[i+1],label = 'MOD:'+model_name_list[i+1])     
+        plt.title('Comparison at '+str(station_name[0])+' on '+str(release_time)+' UTC')
+        plt.legend()
+        plt.ylim(altitude_range[0],altitude_range[1])
+        plt.xlim(ozone_range[0],ozone_range[1])
+        plt.ylabel('ALT-ground level (km)')
+        plt.xlabel('O3 (ppbv)')
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:76]: this is done!

File: melodies_monet/plots/ozone_sonder_plots.py (hunk 5 above)
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:157]: this is done!

File: melodies_monet/plots/ozone_sonder_plots.py (hunk 8 above)
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:230]: this is done!

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        #add a line plot
+        ax.plot(O3_OBS,ALT_sl,'-k*',label = 'OBS:'+model_name_list[0])
+
+        for i in range(len(O3_MODEL_ALL)):
+            output_list_model = split_by_threshold(O3_MODEL_ALL[i],ALT_sl,altitude_threshold_list)
+            bplot_model=ax.boxplot(output_list_model,vert = False,**boxplot_kwargs,positions=position_list_mid)
+            #add a line plot
+            ax.plot(O3_MODEL_ALL[i],ALT_sl,'-*',color=fill_color_list[i+1],label = 'MODEL:'+model_name_list[i+1])
+
+            colors=[fill_color_list[i+1]]*(len(altitude_threshold_list)-1)
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:148]: this is done!

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+        f,ax = plt.subplots(figsize=(8,8))
+
+    #set default text size
+    def_text = dict(fontsize=20)
+    if text_dict is not None:
+        text_kwargs = {**def_text, **text_dict}
+    else:
+        text_kwargs = def_text
+
+    #set color style
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:112]: this is done!

File: melodies_monet/plots/ozone_sonder_plots.py [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+    obs_dict = dict(color='k', linestyle='-',marker='*', linewidth=2.0, markersize=5.)
+    if plot_dict is not None:
+        plot_kwargs = {**obs_dict, **plot_dict}
+        plot_kwargs.pop('column')
+    else:
+        plot_kwargs = obs_dict
+
```
[the remark this answers is not in the document]
Comment (btang1) [reply melodies_monet/plots/ozone_sonder_plots.py:119]: this is done!

**Comment by zmoon** [comment]
@rschwant we have an example now but only in YAML (runs with `melodies-monet run`). We could upload the figures and create a docs page for it, to display the figures. I could also upload the UFS-AQM data (unfortunately couldn't use your file, as the time didn't overlap with any of the GML ozonesondes) for the example data loading mechanism, like you did.

**Review by rschwant — COMMENTED** [review]
Great job addressing my comments so far!

I just have a question about setting up a jupyter notebook to connect your example to the ReadTheDocs completely.

See these here:
https://github.com/NOAA-CSL/MELODIES-MONET/tree/develop/docs/examples

Is there a missing commit maybe for this?

And then in your yaml file you comment out the code for an aggregate_plot_os, but since this is not available yet in the code can you just delete these lines instead just for clarity. Also the first line of the yaml file is missing a G in General.

Thanks, and we can talk about this in our upcoming meeting too!

**Comment by DWesl** [comment]
I'm trying to use this branch for stratospheric ozone and keep running into problems resulting from the model data being on pressure levels with a 1-D coordinate variable for model pressure.  Is this an expected use-case, and am I missing something simple?

**Comment by btang1** [comment]
> I'm trying to use this branch for stratospheric ozone and keep running into problems resulting from the model data being on pressure levels with a 1-D coordinate variable for model pressure. Is this an expected use-case, and am I missing something simple?
I am NOT exactly can solve you question. but we can look into some details of my code to make it more clear to be use.

1) the z-axis is based on pressure level
2) we designed and tested for surface to 20km around. so it should cover stratosphere. 
3) make sure for model you have more than 1 input file, this code will do time interpolation from closest 2 time step.

hope that helps.

and by the way, what observation data and model data you used?

**Comment by DWesl** [comment]
> > I'm trying to use this branch for stratospheric ozone and keep running into problems resulting from the model data being on pressure levels with a 1-D coordinate variable for model pressure. Is this an expected use-case, and am I missing something simple?
>
> I am NOT exactly can solve your question. but we can look into some details of my code to make it more clear to be use.
> 
> 1. the z-axis is based on pressure level
> 2. we designed and tested for surface to 20km around. so it should cover stratosphere.
> 3. make sure for model you have more than 1 input file, this code will do time interpolation from closest 2 time step.
> 
> hope that helps.
> 
> and by the way, what observation data and model data you used?

Post-processed GFS O3MR 1-degree grib2 files.  My specific example is from my own run, but [the operational output](https://nomads.ncep.noaa.gov/cgi-bin/filter_gfs_1p00.pl?dir=%2Fgfs.20241104%2F18%2Fatmos&file=gfs.t18z.pgrb2.1p00.f006&var_O3MR=on&lev_300_mb=on&lev_250_mb=on&lev_200_mb=on&lev_150_mb=on&lev_100_mb=on&lev_70_mb=on&lev_50_mb=on&lev_40_mb=on&lev_30_mb=on&lev_20_mb=on&lev_15_mb=on&lev_10_mb=on&lev_7_mb=on&lev_5_mb=on&lev_3_mb=on&lev_2_mb=on&lev_1_mb=on) should be close.  I ran it through [noaa-oar-arl/fv3grib2nc4](https://github.com/noaa-oar-arl/fv3grib2nc4), tried to follow examples for the configuration file, and ran into errors relating to `pressure_model` being a 1-D coordinate variable instead of a 3-D data variable.

**Comment by DWesl** [comment]
That script seems to be a wrapper around `wgrib2`, which produces only a 1-D pressure variable when the data is on pressure levels.

From your description, the solution would seem to be adding a variable `pressure_model` to the output file that is horizontally constant?

**Comment by btang1** [comment]
> That script seems to be a wrapper around `wgrib2`, which produces only a 1-D pressure variable when the data is on pressure levels.
> 
> From your description, the solution would seem to be adding a variable `pressure_model` to the output file that is horizontally constant?

yes. let is try that. and the pressure variable horizontal dimension need to match model horizontal dimension, try to see if that works.

**Comment by DWesl** [comment]
Okay, will try adding something like this to fv3grib2nc4 and run again
```python
data = xr.open_dataset(out_file)
pressure_model, o3mr = xr.broadcast(data.coords["plevel"], data["O3MR"].isel(time=0))
pressure_model.name = "pressure_model"
pressure_model.to_netcdf(out_file, mode="a")
```
Thanks for the ideas

**Comment by btang1** [comment]
> Okay, will try adding something like this to fv3grib2nc4 and run again
> 
> ```python
> data = xr.open_dataset(out_file)
> pressure_model, o3mr = xr.broadcast(data.coords["plevel"], data["O3MR"].isel(time=0))
> pressure_model.name = "pressure_model"
> pressure_model.to_netcdf(out_file, mode="a")
> ```
> 
> Thanks for the ideas

Any updates? is adding 3d pressure working for your case?

**Comment by DWesl** [comment]
3D pressure got the axes in the wrong order, so the vertical interpolation failed.
4D pressure got through vertical interpolation, and is now complaining that it can't add the old pressure coordinate to the new dataset.

I'm going to check whether renaming the current pressure coordinate `plevel` to `z` helps any.  I'll need to change the loop where the renaming happens, but that's straightforward.

**Comment by btang1** [comment]
@zmoon looks like you finished the requirement @rschwant asked.
could we merge this now? @rschwant @zmoon

**Comment by rschwant** [comment]
Yes, I can approve this. @zmoon do you know why the docs are failing now? I think this is independent of the changes here.

**Comment by zmoon** [comment]
@rschwant there was a bug, which the CI caught for us, I think I've fixed it.

**Comment by btang1** [comment]
> Great, glad the CI found the bug. Already very useful! This is approved! Glad this is getting into the tool.

thanks! @rschwant @zmoon it has been a long and wonderful journey!

I saw it been approved. has it been merged to develop yet?

**Pull request merged**